RecoTracker/MkFit/plugins/MkFitProducer.cc

```diff
@@ -175,9 +183,22 @@ void MkFitProducer::produce(edm::StreamID iID, edm::Event& iEvent, const edm::Ev
   auto seeds_mutable = seeds.seeds();
   mkfit::TrackVec tracks;
 
+  if (not config_) {
+    // TODO: strong assumption that a reference to an IterationConfig
+    // is valid while the ConfigJson_Load_File is called in other
+    // thread (e.g. two MkFitProducer instances do not use the same
+    // IterationConfig)
+    std::scoped_lock lk{configMutex};
+    if (not config_) {
+      // TODO: gross violation of policies for EventSetup products
+      config_ =
+          &mkfit::ConfigJson_Load_File(const_cast<mkfit::IterationsInfo&>(mkFitGeom.iterationsInfo()), configFile_);
```

Comment:
This shortcut must be addressed. The way to go depends on near/mid term plans for code evolution. @osschar How much would your suggestion https://github.com/trackreco/mkFit/pull/315#issuecomment-839460292 impact here? (after reading it again I'd assume "not much")

Reply:
I don't understand what is the problem :) const_cast?

My suggestion was to do initialization internally in cmssw, then you get IterationsInfo. This could be kept internal.

The Load function can then access this, possibly as const, as you can copy the default to a new ItertaionConfig and load over this new copy which is then fully owned by the caller.

Reply:
I was trying to provide you "building blocks" so you can then do the kosher thing :)

Reply:
The particular problem here is a modification of an EventSetup product after it has been created by and ESProducer. While waiting for
> In the future we then cleanup this code so IterationConfigs can be created independently, per iteration.

it seems to me that we can either
1. Go back to reading one JSON file in the `MkFitGeometryESProducer` and using magic numbers to identify iterations in the per-iteration python configuration files (or provide a way to map iteration names to the internal index), or
2. `MkFitProducer` makes a local copy of `IterationsInfo` (that would be a copy per iteration times the number of EDM streams) and calls `ConfigJson_Load_File` on that

I'm perhaps leaning towards option 2 because then the configuration structure would stay the same as here (assuming this is the one intended for this round of integration).

@slava77 @osschar Any further thoughts?

Reply:
why not have a separate ES product (and a separate ES producer instance) per file/iteration?

Reply:
i didn't think about that. 
Just to invent a reason: some analyzer may use it.

Reply:
(I wrote my previous message too hastily)

> why not have a separate ES product (and a separate ES producer instance) per file/iteration?

That would be a good way to avoid the per-stream copies of the `IterationConfig` (that would be needed even with "`IterationConfig`s can be created independently per iteration"). It would not help though for the immediate need to copy `IterationsInfo` per iteration though, but would help to avoid the per-stream copies of that too.